### ISSUE TITLE
anchor_labeler: batch_label_anchors issue with single bounding box

### ISSUE THREAD

**Issue by heye0507** [rwightman/efficientdet-pytorch#44]
Hi @rwightman 

Thank you so much for making this repo.

I am currently experiencing an issue calling batch_label_anchors when ground truth bounding box list only has 1 bbox. Not sure what might caused this issue, I am wondering if you can take a look. Thanks in advance :)

Issue:
anchor_labeler.batch_label_anchors () has index out of range error. Trace attached

Setup anchor:
```
model_config = get_efficientdet_config('tf_efficientdet_d0')
model = EfficientDet(model_config,pretrained_backbone=True)
model_config.num_classes = 1
model_config.image_size = 512

anchors = Anchors(
    model_config.min_level,model_config.max_level,
    model_config.num_scales, model_config.aspect_ratios,
    model_config.anchor_scale, model_config.image_size
    )


anchor_labeler = AnchorLabeler(anchors,model_config.num_classes,match_threshold=0.5)
```

Reproduce:
```
tb = torch.tensor([[468.,353.,52.,386.5]])
tb = tb.int().float()
tlbl = torch.tensor([1.])
cls_targets, box_targets,num_positives = anchor_labeler.batch_label_anchors(1,[tb],[tlbl])
```

Trace:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-45-e8bceaf11fb2> in <module>
----> 1 cls_targets, box_targets,num_positives = anchor_labeler.batch_label_anchors(1,[tb],[tlbl])

/opt/conda/envs/fastai2/lib/python3.7/site-packages/effdet/anchors.py in batch_label_anchors(self, batch_size, gt_boxes, gt_classes)
    394             # cls_weights, box_weights are not used
    395             cls_targets, _, box_targets, _, matches = self.target_assigner.assign(
--> 396                 anchor_box_list, BoxList(gt_boxes[i]), gt_classes[i])
    397 
    398             # class labels start from 1 and the background class = -1

/opt/conda/envs/fastai2/lib/python3.7/site-packages/effdet/object_detection/target_assigner.py in assign(self, anchors, groundtruth_boxes, groundtruth_labels, groundtruth_weights)
    144         match_quality_matrix = self._similarity_calc.compare(groundtruth_boxes, anchors)
    145         match = self._matcher.match(match_quality_matrix)
--> 146         reg_targets = self._create_regression_targets(anchors, groundtruth_boxes, match)
    147         cls_targets = self._create_classification_targets(groundtruth_labels, match)
    148         reg_weights = self._create_regression_weights(match, groundtruth_weights)

/opt/conda/envs/fastai2/lib/python3.7/site-packages/effdet/object_detection/target_assigner.py in _create_regression_targets(self, anchors, groundtruth_boxes, match)
    167         zero_box = torch.zeros(4, device=device)
    168         matched_gt_boxes = match.gather_based_on_match(
--> 169             groundtruth_boxes.boxes(), unmatched_value=zero_box, ignored_value=zero_box)
    170         matched_gt_boxlist = box_list.BoxList(matched_gt_boxes)
    171         if groundtruth_boxes.has_field(self._keypoints_field_name):

/opt/conda/envs/fastai2/lib/python3.7/site-packages/effdet/object_detection/matcher.py in gather_based_on_match(self, input_tensor, unmatched_value, ignored_value)
    171         input_tensor = torch.cat([ss, input_tensor], dim=0)
    172         gather_indices = torch.clamp(self.match_results + 2, min=0)
--> 173         gathered_tensor = torch.index_select(input_tensor, 0, gather_indices)
    174         return gathered_tensor

IndexError: index out of range in self

Here are some values:
ipdb> p input_tensor
tensor([[  0.,   0.,   0.,   0.],
        [  0.,   0.,   0.,   0.],
        [468., 353.,  52., 386.]])

ipdb> p gather_indices.shape
torch.Size([49104])

ipdb> p gather_indices
tensor([    1,     1,     1,  ...,     1,     1, 24554])

ipdb> p self.match_results
tensor([   -1,    -1,    -1,  ...,    -1,    -1, 24552])

```

**Comment by rwightman** [comment]
@heye0507 thanks for repro, wonder if a dim is benig dropped somewhere in the n=1 case, hmm

**Comment by lgvaz** [comment]
Any updates on this? I believe I'm facing the same error:

```python
self = <effdet.object_detection.matcher.Match object at 0x7f27110ba210>
input_tensor = tensor([[  0.0000,   0.0000,   0.0000,   0.0000],
        [  0.0000,   0.0000,   0.0000,   0.0000],
        [188.3483,  88.2405, 385.1532, 447.3587]])
unmatched_value = tensor([0., 0., 0., 0.])
ignored_value = tensor([0., 0., 0., 0.])

    def gather_based_on_match(self, input_tensor, unmatched_value, ignored_value):
        ss = torch.stack([ignored_value, unmatched_value])
        input_tensor = torch.cat([ss, input_tensor], dim=0)
        gather_indices = torch.clamp(self.match_results + 2, min=0)
>       gathered_tensor = torch.index_select(input_tensor, 0, gather_indices)
E       IndexError: index out of range in self
```

**Comment by lgvaz** [comment]
I don't know if this is the cause, but I get the same error if  `match_results` has a value > 2:

```python
match_results = tensor([3])

input_tensor = tensor(
    [
        [0.0000, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000, 0.0000],
        [188.3483, 88.2405, 385.1532, 447.3587],
    ]
)
unmatched_value = tensor([0.0, 0.0, 0.0, 0.0])
ignored_value = tensor([0.0, 0.0, 0.0, 0.0])

ss = torch.stack([ignored_value, unmatched_value])

input_tensor = torch.cat([ss, input_tensor], dim=0)
gather_indices = torch.clamp(tensor(match_results) + 2, min=0)

gathered_tensor = torch.index_select(input_tensor, 0, gather_indices)
```

```python
    gathered_tensor = torch.index_select(input_tensor, 0, gather_indices)
IndexError: index out of range in self
```

**Comment by heye0507** [comment]
Hi @rwightman, @lgvaz 

Thanks for the response.  Sorry for the late reply, was pulled to another project forgot to comment back.

Two things I want to confirm here, 

1. Using Bench train everything works pretty well. (I was in the kaggle wheat detection, pretty sure you are aware of it :) ) I have tested Eff-det0, Eff-det5 and Eff-det6, things working very well.

2.  The issue reported here seems to related to specific data / frame work you are using. I was porting Effdet to Fastai2, and randomly encountered this issue. (I later refactored my code and the issue seems gone, apologize that I still didn't find root cause of it).

But here is one of my testing example, if you want to take a look. (I do have a better version that I packed everything into a Fastai callback class, but I forgot to push it to github).

[Running in Fastai2](https://github.com/heye0507/wheat_detection/blob/master/nbs/004_effdet_run_test.ipynb)

Best,

**Comment by lgvaz** [comment]
TL;DR: The error seems to be specific to CPU

---
Hello @heye0507, thanks for the update!

Funnily I was getting this error only on my unittests and not when actually training the model. which got me to the question: WHY?

The only difference is that my unittests are run on CPU, moving the tests to GPU **"solves the issue"**

I found [this](https://github.com/pytorch/pytorch/issues/571) old issue (from 2017) about `index_select` not returning out of bound of errors on GPU. The issue is closed, but it's not very clear if the issue was fixed.

@rwightman Is it possible that we have a bug in `Match` that is being shadowed by this?

**Comment by rwightman** [comment]
@lgvaz certainly possible there is a bug or corner case that's being missed on the GPU, back in the day there were a number of ops that didn't check bounds or error conditions well on the GPU, thought that was all fixed. 

Also possible that the input is not valid. What's the test case that triggers this? Have you tried training with COCO data on the CPU (for a bit). @heye0507 original test case would never happen in reality, batch size of 1 is not an input you'd ever see, batchnorm doesn't work in train mode with batch size 1. Also, 1 bounding box at the point it's fed to the network is not something you'd typicall see. 

The whole batch needs to have the same box instances dimension, fixing that to 1 means your ground truth targets can have at most 1 bbox for all images. While not impossible to imagine a dataset where you have one box or not, doesn't seem that common and if indeed it is a dim size == 1 issue, you coudl set max instances to 2 and not see any performance impact.

All of this anchor code was adapted by me from the Google autml/efficientdet code base, which took the anchor code from the  official retinanet tensorflow impl. I converted it from TF 1 graph ops into Pytorch eager in a couple of evenings with a glass (or two) of scotch. Ran through a debugging evening where I fixed issues mostly in the gather -> gather/select ops (the most different) and then a few performance adjustments later. The original code and my own adapation are certainly NOT defensive, there are not many checks so making sure the input is valid is quite important. I'm hoping to find a chunk of time to look into the IOU PR and possible improvements, perhaps I can get normal training running on the CPU and do some checks of this while I'm fiddling....

**Comment by rwightman** [comment]
Interesting, from tf.gather that I use torch.index_select, 

https://www.tensorflow.org/api_docs/python/tf/gather `Note that on CPU, if an out of bound index is found, an error is returned. On GPU, if an out of bound index is found, a 0 is stored in the corresponding output value`

Possibility also that the original code has some issues that are ignored for GPU (and likely TPU).

**Comment by lgvaz** [comment]
Instead of trying to train COCO on cpu maybe you can try [this](https://cvbp.blob.core.windows.net/public/datasets/object_detection/odFridgeObjects.zip) dataset? It's very small and we can be sure the error will pop up very quickly.

The dataset is in VOC format, but there are a lot of tools that convert voc to coco.

I believe the error should appear regardeless of the batch size, I still don't understand **why** the error is happening still, any thoughts on what might be happening?

---
On a side note, I laughed quite hard at:
>  in a couple of evenings with a glass (or two) of scotch.  
Convert TF graphs to pytorch really doesn't sound pleasant hahahahaha

**Comment by Chris-hughes10** [comment]
I can confirm I am also being bitten by this error in the case where I have a single bounding box, irrespective of batch size, running on cpu.

**Comment by rwightman** [comment]
@Chris-hughes10 @lgvaz  to be clear, when one says 'single bounding box', does that mean a single valid bounding box in an N instance batched tensor, or that your bbox/cls tensors only have a dim of 1 for the number of instances? If it's the latter, what happens if you change the dim to say 2?

**Comment by Chris-hughes10** [comment]
Hi @rwightman, by single bounding box I mean a Nx4 tensor; one ground truth bbox per image. When I used an image with two ground truth boxes I didn't hit the problem.

**Comment by world2vec** [comment]
@rwightman @heye0507 
This appears to be bug of torch.argmax
`
torch.argmax(torch.tensor(np.random.rand(1, 30000)), 0)[-1]
Out[36]: tensor(0)
torch.argmax(torch.tensor(np.random.rand(1, 40000)), 0)[-1]
Out[37]: tensor(30000) # should always be 0

torch.argmax(torch.tensor(np.random.rand(2, 40000)), 0)[-1]
Out[38]: tensor(1)
torch.argmax(torch.tensor(np.random.rand(2, 40000)), 0)
Out[39]: tensor([1, 0, 0,  ..., 1, 0, 0])
`

**Comment by rwightman** [comment]
@world2vec weird, I can't reproduce that, but I did notice that the torch argmax behaviour is not consistent with numpy when values are the same, torch cpu returns the highest same index, GPU seems rather arbitrary. That was apparently just fixed this month.

**Comment by world2vec** [comment]
@rwightman 
`torch.__version__
Out[3]: '1.5.0+cu101'`
kindly be noted CUDA was disabled.

You are right,  loooks 1.6 has solved the bug:
```In [1]: import torch

In [2]: torch.cuda.is_available()
Out[2]: False

In [3]: torch.__version__
Out[3]: '1.6.0+cu101'

In [4]: torch.argmax(torch.tensor(np.random.rand(1, 40000)), 0)[-1]
Out[4]: tensor(0)
```


Here is my temporary solution.

```
git diff argmax_matcher.py
diff --git a/effdet/object_detection/argmax_matcher.py b/effdet/object_detection/argmax_matcher.py
index 2b8d9b1..c73fc3b 100644
--- a/effdet/object_detection/argmax_matcher.py
+++ b/effdet/object_detection/argmax_matcher.py
@@ -115,7 +115,10 @@ class ArgMaxMatcher(object):  # cannot inherit with torchscript
             matches:  int32 tensor indicating the row each column matches to.
         """
         # Matches for each column
-        matches = torch.argmax(similarity_matrix, 0)
+        if similarity_matrix.shape[0] == 1:
+            matches = torch.tensor([0]*similarity_matrix.shape[1])
+        else:
+            matches = torch.argmax(similarity_matrix, 0)
 
         # Deal with matched and unmatched threshold
         if self._matched_threshold is not None:
@@ -135,7 +138,11 @@ class ArgMaxMatcher(object):  # cannot inherit with torchscript
         if self._force_match_for_each_row:
             force_match_column_ids = torch.argmax(similarity_matrix, 1)
             force_match_column_indicators = one_hot(force_match_column_ids, similarity_matrix.shape[1])
-            force_match_row_ids = torch.argmax(force_match_column_indicators, 0)
+
+            if force_match_column_indicators.shape[0] == 1:
+                force_match_row_ids = torch.tensor([0]*force_match_column_indicators.shape[1])
+            else:
+                force_match_row_ids = torch.argmax(force_match_column_indicators, 0)
             force_match_column_mask = torch.max(force_match_column_indicators, 0)[0] != 0
             final_matches = torch.where(force_match_column_mask, force_match_row_ids, matches)
             return final_matches

```

**Comment by rwightman** [comment]
Interesting, I've almost exclusively used this codebase with PyTorch 1.4 and 1.6. 1.5 had some big performance regerssions for the JIT scripting fusion that I use for the Swish activation and some of the post-processing for these models. Wonder if everyone having these issues like this using 1.5?

**Comment by rwightman** [comment]
@Chris-hughes10 @lgvaz @heye0507 Can anyone confirm if this issue happens on PyTorch 1.6? I'm going to ignore any problems with 1.5 due to the issue @world2vec mentioned (that I have confirmed).

**Comment by bguan** [comment]
I can confirm that I am running into same issue with effdet 0.1.6, under Python ver 3.8.2, torch 1.6.0, torchvision 0.7.0, fastai 2.0.13, pytorch_lightning 0.9.1rc3, icevision 0.1.5

```
ValueError                                Traceback (most recent call last)
/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _with_events(self, f, event_type, ex, final)
    154     def _with_events(self, f, event_type, ex, final=noop):
--> 155         try:       self(f'before_{event_type}')       ;f()
    156         except ex: self(f'after_cancel_{event_type}')

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _do_epoch(self)
    190     def _do_epoch(self):
--> 191         self._do_epoch_train()
    192         self._do_epoch_validate()

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _do_epoch_train(self)
    182         self.dl = self.dls.train
--> 183         self._with_events(self.all_batches, 'train', CancelTrainException)
    184 

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _with_events(self, f, event_type, ex, final)
    154     def _with_events(self, f, event_type, ex, final=noop):
--> 155         try:       self(f'before_{event_type}')       ;f()
    156         except ex: self(f'after_cancel_{event_type}')

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in all_batches(self)
    160         self.n_iter = len(self.dl)
--> 161         for o in enumerate(self.dl): self.one_batch(*o)
    162 

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in one_batch(self, i, b)
    178         self._split(b)
--> 179         self._with_events(self._do_one_batch, 'batch', CancelBatchException)
    180 

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _with_events(self, f, event_type, ex, final)
    154     def _with_events(self, f, event_type, ex, final=noop):
--> 155         try:       self(f'before_{event_type}')       ;f()
    156         except ex: self(f'after_cancel_{event_type}')

/usr/local/lib/python3.8/dist-packages/fastai/learner.py in _do_one_batch(self)
    163     def _do_one_batch(self):
--> 164         self.pred = self.model(*self.xb)
    165         self('after_pred')

~/.local/lib/python3.8/site-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(

~/.local/lib/python3.8/site-packages/effdet/bench.py in forward(self, x, target)
     92         class_out, box_out = self.model(x)
---> 93         cls_targets, box_targets, num_positives = self.anchor_labeler.batch_label_anchors(
     94             x.shape[0], target['bbox'], target['cls'])

~/.local/lib/python3.8/site-packages/effdet/anchors.py in batch_label_anchors(self, batch_size, gt_boxes, gt_classes)
    410             cls_targets, _, box_targets, _, matches = self.target_assigner.assign(
--> 411                 anchor_box_list, BoxList(gt_boxes[i]), gt_classes[i])
    412 

~/.local/lib/python3.8/site-packages/effdet/object_detection/box_list.py in __init__(self, boxes)
     53         if len(boxes.shape) != 2 or boxes.shape[-1] != 4:
---> 54             raise ValueError('Invalid dimensions for box data.')
     55         if boxes.dtype != torch.float32:

ValueError: Invalid dimensions for box data.
```

I can bypass it by hacking my pipeline to insert a bogus extra box when there is only 1 labeled box on an image.

**Comment by rwightman** [comment]
@bguan I think that's a different, expected issue wrt to the tensor format for the boxes that's not specific to CPU/GPU.

**Comment by bguan** [comment]
@rwightman can you advice how to go about doing it the right way?

**Comment by lgvaz** [comment]
@bguan I think your issue is more related to #71 , if you invoke a debugger and check the value of `boxes` I think you will find an empty tensor.

@rwightman I'm still a bit confused on the resolution of #71, what should be passed as targets if no bboxes and present on the image?

Just to be clearer on that, let's say I have 2 images on my batch, one of then has 2 bboxes and the other has 0, what I'm currently feeding as targets for the model is:

```python
targets['bbox'] = [tensor([[10, 10, 20, 20], [5, 5, 15, 15]]), tensor([])]
```

This last `tensor([])` is what is causing the problem, what should the target be in that case?

For faster_rcnn I'm doing:
```
target["labels"] = torch.zeros(0, dtype=torch.int64)
target["boxes"] = torch.zeros((0, 4), dtype=torch.float32)
```

Will that work here as well?

**Comment by rwightman** [comment]
@bguan @lgvaz the target bbox tensors passed to the bench by my code are shape [B, N, 4] ... where B is the batch size, and N is the number of detection instances. classes are [B, N] for scalar class targets. The bench currently iterates over the batch dimension and passes bbox [N, 4], and class targets [N] to the anchor labeller. For the boxes you cannot strip the first dimension N in the case where there is one (or no boxes).

As per the TF implementation I fix the size of N for all samples regardless of the number of actual boxes. This is done in my collate fn for efficiency. So by default for my training routine I pass one bbox tensor and one class tensor for the batch of size [B, 100, 4], and [B, 100] as 100 is the default max number of instances for coco. I tried setting N to 1 for my batch tensors in the past and it worked fine.

For the fixed N instances, I currently pad the unused instances with zeros. However, in taking another pass over the original code recently I noticed the original looks like it pads out boxes and classes to -1, which ends up as -2 for the classes after subtracting the background, and -2 is treated as a different sentinel value than -1 (background) by the loss/anchor/match code. I've had good results as is, but I'm trying runs with -1 padding to see if it has an impact.

Making targets a list of tensors as in your example, with differing N... well I've certainly never tried that. I can't say it wouldn't/shouldn't work for the anchor/matching code. An empty dim [] tensor certainly won't. (0, 4) might. I never thought about this approach as it seems less efficient, you'll be moving more tensors separately to the GPU, should be less latency to move it in fewer transfers.

With fixed size N dimension padding is required, 

Either way, you definitely cannot pass boxes that don't have a leading N dimension.
